### PR TITLE
The birth date can be extracted also from bis numbers

### PR DESCRIPTION
**Pull request by shkurata**
It will be a good feature if it accepts also BIS numbers

### CODE REVIEW

**Comment by moeriki** [comment]
What is a BIS number?

**Comment by shkurata** [comment]
Een eerste groep van zes cijfers, gevormd door de geboortedatum in de volgorde: jaar, maand, dag. Maand en/of dag kunnen nul zijn indien de exacte geboortedatum niet bekend is. Indien de persoon niet ingeschreven is in het Rijksregister, maar er toch gegevens moeten worden bijgehouden voor de sociale zekerheid, bijvoorbeeld buitenlandse werknemers die minder dan drie maanden in België verblijven of grensarbeiders, dan wordt een bisnummer toegekend. Bij het bisnummer wordt de geboortemaand verhoogd met 20 of 40. Als bij de aanvraag het geslacht bekend is, wordt de geboortemaand verhoogd met 40, anders wordt ze verhoogd met 20. Indien de persoon vluchteling is en de geboortedatum niet gekend is, wordt de geboortemaand op 00 gezet en de geboortedag op 00 gezet.
https://nl.m.wikipedia.org/wiki/Rijksregisternummer

File: src/index.spec.ts [hunk 1]
```diff
@@ -34,7 +34,14 @@ describe('getBirthDate()', () => {
       parseDate('2001-08-14').getTime(),
     );
   });
-
+  it('should extract the correct birthdate from a NRN', () => {
```
Comment (moeriki) [inline src/index.spec.ts:38]: Can you pad an empty line before and after this test like the rest of the file.

File: src/index.ts [hunk 2]
```diff
@@ -36,8 +36,8 @@ export function getAge(
 export function getBirthDate(nrn: NrnInput): Date {
   const { birthDate, serial, checksum } = parse(nrn);
   const partialYear = birthDate.slice(0, 2); // Eg. '86' from '860814'
-  const month = Number(birthDate.slice(2, 4)); // Eg. 8 from '860814'
-  const day = Number(birthDate.slice(4, 6)); // Eg. 14 from '860814'
+  const month = normalizeMonth(birthDate.slice(2, 4)); // Eg. 8 from '860814'
+  const day = Number(birthDate.slice(4, 6)) || 1; // Eg. 14 from '860814'
```
Comment (moeriki) [inline src/index.ts]: Why `|| 1`?

As I understood it some people can have `00` as birth month. I think it's better to throw an error in that case. Returning a `Date` for them would be ambiguous.

File: src/index.ts [hunk 3]
```diff
@@ -98,3 +97,17 @@ export function parse(nrnInput: NrnInput): Nrn {
...
+    if (m === 0) {
+      return 6;
+    }
+    // normal month => normal RNR
+    return m;
+  } else {
+    // BIS number, substract 20 or 40 until you have a valid month
+    return m - 20 < 13 ? m - 20 : m - 40;
+  }
+}
```
Comment (moeriki) [inline src/index.ts:169]: Let's not have this in a separate function. We will need to re-use its logic many times though.

**Comment by moeriki** [comment]
Just FYI. It's better to open an issue first so we can discuss things.

If we are going to add BIS numbers we want to add also some additional functions in regard. I propose the following.

`isBisNumber()`: returns `true` if birth month was increased with `20` or `40`.
`isBisBirthdateKnown()` returns `true` if birth month `>0` after subtracting `20` or `40`.
`isBisGenderKnown()` returns `true` if birth month was increased with `40`.
`isNrnNumber()`: returns `true` if birth month was not increased with `20` or `40`.

Additionally I think the following changes are in order.

`getBirthDate()` should throw an Error if the birth date is unknown. We can't be ambiguous about this.

**Comment by moeriki** [comment]
We have an return interface on `parse()` we should probably update.

https://github.com/Moeriki/be-nrn/blob/master/src/index.ts#L8

```ts
export interface Nrn {
  birthDate: readonly [string, string, string];
  serial: string;
  checksum: string;
}
```

So we can include the slicing of the date parts in the `parse` already.

**Comment by moeriki** [comment]
It doesn't seem like you ran linting. Make sure to run prettier when you are done.

**Review by moeriki — CHANGES_REQUESTED** [review]

File: src/index.ts (hunk 2 above)
Comment (shkurata) [reply src/index.ts]: Some people can have 00 for the birth month and also 00 for the birth date, so we decided to use the middle of the year and  the first of the month.

**Comment by shkurata** [comment]
We had to find a fast solution, because we had an issue in PROD.

File: src/index.ts (hunk 2 above)
Comment (moeriki) [reply src/index.ts]: Yeah so I don't want that. You can't tell the difference between known and unknown dates. Just throw an error and resolve it in your own application. It doesn't make sense for the library to make these kinds of defaults.

**Comment by moeriki** [comment]
Merged and optimised.

I renamed a few methods.

`isBisBirthdateKnown` -> `isBirthDateKnown`
`isBisGenderKnown` -> `isGenderKnown`

They now work regardless of BIS / NRN.

I removed `isValidNrnNumber` because this is an opinionated implementation. Being "valid" is not something I aim to solve in this library.

I fixed one bug in your implementation where `isGenderKnown()` returned `false` if the gender was known but the date of birth was not.

```js
    expect(nrnUtils.isGenderKnown('814000 896 60')).toBe(true);
```

This is published a `be-nrn@1.1.2`.

**Pull request merged**